extensions-core/lookups-cached-global/src/main/java/io/druid/server/lookup/namespace/cache/CacheScheduler.java

```diff
@@ -0,0 +1,487 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.server.lookup.namespace.cache;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Throwables;
+import com.google.inject.Inject;
+import com.metamx.emitter.service.ServiceEmitter;
+import com.metamx.emitter.service.ServiceMetricEvent;
+import io.druid.java.util.common.ISE;
+import io.druid.java.util.common.logger.Logger;
+import io.druid.query.lookup.namespace.ExtractionNamespace;
+import io.druid.query.lookup.namespace.ExtractionNamespaceCacheFactory;
+import sun.misc.Cleaner;
+
+import javax.annotation.Nullable;
+import java.util.Map;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Usage:
+ * <pre>{@code
+ * CacheScheduler.Entry entry = cacheScheduler.schedule(namespace); // or scheduleAndWait(namespace, timeout)
+ * CacheState cacheState = entry.getCacheState();
+ * if (cacheState instanceof NoCache) {
+ *   // the cache is not yet created, or already disposed
+ * } else if (cacheState instanceof VersionedCache) {
+ *   Map<String, String> cache = ((VersionedCache) cacheState).getCache(); // use the cache
+ * }
+ * ...
+ * entry.close(); // dispose the last VersionedCache and unschedule future updates
+ * }</pre>
+ */
+public final class CacheScheduler
+{
+  private static final Logger log = new Logger(CacheScheduler.class);
+
+  public final class Entry<T extends ExtractionNamespace> implements AutoCloseable
+  {
+    private final EntryImpl<T> impl;
+
+    private Entry(final T namespace, final ExtractionNamespaceCacheFactory<T> cachePopulator)
+    {
+      impl = new EntryImpl<>(namespace, this, cachePopulator);
+    }
+
+    /**
+     * Returns the last cache state, either {@link NoCache} or {@link VersionedCache}.
+     */
+    public CacheState getCacheState()
+    {
+      return impl.cacheStateHolder.get();
+    }
+
+    /**
+     * @return the entry's cache if it is already initialized and not yet disposed
+     * @throws IllegalStateException if the entry's cache is not yet initialized, or {@link #close()} has
+     * already been called
+     */
+    public Map<String, String> getCache()
+    {
+      CacheState cacheState = getCacheState();
+      if (cacheState instanceof VersionedCache) {
+        return ((VersionedCache) cacheState).getCache();
+      } else {
+        throw new IllegalStateException("Cannot get cache: " + cacheState);
```

Comment:
Please `io.druid.java.util.common.ISE` for `IllegalStateException`, which under the hood uses `StringUtils.safeFormat` for error message printing.